packages/wrangler/src/package-manager.ts

```diff
@@ -0,0 +1,88 @@
+import { existsSync } from "node:fs";
+import { join } from "node:path";
+import { execa, execaCommandSync } from "execa";
+
+export interface PackageManager {
+  addDevDeps(...packages: string[]): Promise<void>;
+  install(): Promise<void>;
+}
+
+export async function getPackageManager(root: string): Promise<PackageManager> {
+  const [hasYarn, hasNpm] = await Promise.all([supportsYarn(), supportsNpm()]);
+  const hasYarnLock = existsSync(join(root, "yarn.lock"));
+  const hasNpmLock = existsSync(join(root, "package-lock.json"));
+
+  if (hasYarn && hasYarnLock && !hasNpmLock) {
+    return new YarnPackageManager();
+  } else if (hasNpm && hasNpmLock && !hasYarnLock) {
+    return new NpmPackageManager();
+  } else if (hasYarn && !hasNpm) {
+    return new YarnPackageManager();
+  } else if (hasNpm && !hasYarn) {
+    return new NpmPackageManager();
+  } else {
+    return new NpmPackageManager();
+  }
+}
+
+async function supports(name: string): Promise<boolean> {
+  try {
+    execaCommandSync(`${name} --version`, { stdio: "ignore" });
+    return true;
+  } catch {
+    return false;
+  }
+}
+
+function supportsYarn(): Promise<boolean> {
+  return supports("yarn");
+}
+
+function supportsNpm(): Promise<boolean> {
+  return supports("npm");
+}
+
+/**
+ * Manage packages using npm
+ */
+export class NpmPackageManager implements PackageManager {
```

Comment:
I don't feel strongly about this, but why is this a class? Can it be a regular object like before?

Reply:
Coming from an OO background, a class seemed like the obvious paradigm here.
I am interesting in the argument against using a class for this sort of thing?

Reply:
There's no state, no constructor arguments, and there will only be one instance/singleton required. So there's no visible benefit of using a class here (unless I'm missing something? Does it make types or tests simpler? I could be wrong ofc.)

Reply:
Yes - all good points. And in fact on reflection, using a class actually makes the testing slightly more awkward (see `...prototype` bits. I'll change.